pyteal/ast/abi/util_test.py

```diff
@@ -10,289 +10,350 @@
+            abi.Uint16TypeSpec(),
+            abi.ByteTypeSpec(),
+            abi.BoolTypeSpec(),
+            abi.TupleTypeSpec(),
         ),
-    ]
+    ),
+    TypeAnnotationTest(
+        annotation=abi.Tuple5,
+        expected=TypeError,
+    ),
+    TypeAnnotationTest(
+        annotation=ExampleNamedTuple, expected=ExampleNamedTuple().type_spec()
+    ),
+    TypeAnnotationTest(
+        annotation=list[abi.Uint16],
+        expected=TypeError,
+    ),
+    TypeAnnotationTest(annotation=abi.Transaction, expected=abi.TransactionTypeSpec()),
+    TypeAnnotationTest(
+        annotation=abi.PaymentTransaction, expected=abi.PaymentTransactionTypeSpec()
+    ),
+    TypeAnnotationTest(
+        annotation=abi.KeyRegisterTransaction,
+        expected=abi.KeyRegisterTransactionTypeSpec(),
+    ),
+    TypeAnnotationTest(
+        annotation=abi.AssetConfigTransaction,
+        expected=abi.AssetConfigTransactionTypeSpec(),
+    ),
+    TypeAnnotationTest(
+        annotation=abi.AssetFreezeTransaction,
+        expected=abi.AssetFreezeTransactionTypeSpec(),
+    ),
+    TypeAnnotationTest(
+        annotation=abi.AssetTransferTransaction,
+        expected=abi.AssetTransferTransactionTypeSpec(),
+    ),
+    TypeAnnotationTest(
+        annotation=abi.ApplicationCallTransaction,
+        expected=abi.ApplicationCallTransactionTypeSpec(),
+    ),
+    TypeAnnotationTest(annotation=abi.Account, expected=abi.AccountTypeSpec()),
+    TypeAnnotationTest(annotation=abi.Asset, expected=abi.AssetTypeSpec()),
+    TypeAnnotationTest(annotation=abi.Application, expected=abi.ApplicationTypeSpec()),
+]
 
-    for i, test in enumerate(tests):
-        if not isinstance(test.expected, abi.TypeSpec):
-            with pytest.raises(test.expected):
-                type_spec_from_annotation(test.annotation)
-            continue
 
-        actual = type_spec_from_annotation(test.annotation)
-        assert actual == test.expected, "Test at index {} failed".format(i)
+@pytest.mark.parametrize("annotation, expected", TYPE_ANNOTATION_TEST_CASES)
+def test_type_spec_from_annotation(
+    annotation: Any, expected: abi.TypeSpec | type[Exception]
+):
+    if not isinstance(expected, abi.TypeSpec):
+        with pytest.raises(expected):
+            abi.type_spec_from_annotation(annotation)
+        return
+
+    actual = abi.type_spec_from_annotation(annotation)
+    assert actual == expected
+
+    new_instance = actual.new_instance()
+
+    annotation_origin = get_origin(annotation)
+    if annotation_origin is None:
+        # get_origin will return None for annotations without generic args, e.g. `Byte`
+        annotation_origin = annotation
+
+    assert isinstance(
+        new_instance, annotation_origin
+    ), "TypeSpec.new_instance() returns a value that does not match the annotation type"
+
+    assert (
+        actual == new_instance.type_spec()
+    ), "TypeSpec.new_instance().type_spec() does not match original TypeSpec"
```

Comment:
for other pair of eyes: this part is the roundtrip testing for `TypeSpec`, generated new instance should
- hold the same type-spec as the type-spec generates it, 
- and such new instance is an instance of origin class of annotation